### ISSUE TITLE
Several issues with map editor

### ISSUE THREAD

**Issue by koverhbarc** [sirjuddington/SLADE#739]
**SLADE Version**: 3.1.1.5 and 3.1.0
**Operating System**: Windows XP

**Issue Details**:
SLADE has always hung while closing if I have used the map editor. It enters an infinite loop and must be killed with Task Manager.

For testing purposes, I recently created a version of Planisphere 2 with all Things (except a player start) removed. Yes, I know there's a better way than deleting all 2023, and I finally did it that way, but when I tried I found:

- Can't complete deleting all of them in one go. SLADE, due to a large memory leak, crashes before completion.
- SLADE automatically uncompresses the sidedefs when editing, and doesn't re-compress them while saving. However, this map has over 64k when uncompressed, and thus overflows - all past 65535 are lost. SLADE doesn't even give a warning about this corruption.

I upgraded from 3.1.0 thinking it might fix these things, but it did not; I noticed no change. (But I did have to install a new version of the VC++ Redistributable - could having more than one installed be the cause of the first issue?)

**Comment by sirjuddington** [comment]
Does the crash when deleting all the things happen in the latest 3.1.2 beta? I can't reproduce it there.

SLADE doesn't currently support writing compressed sidedefs, no. It should probably (until that is implemented) show an error and fail to save if there are too many.

**Comment by koverhbarc** [comment]
Yes, it does, and it also occurs on a Windows 7 machine with more memory. The actual crash happens only at the very end, after >1900 things have been deleted and SLADE has leaked 1500 MB (more memory than I've ever seen any program use before!) - that's nearly a megabyte per thing.

Just as importantly, the 3.1.2 build won't run on my XP machine, even though it is supposed to support XP (it gives the generic 'has encountered a problem and needs to close' error message on launch); the 'XP' binaries do run on 7. I transferred them over because the Win7 machine doesn't have 7zip; never needed it for anything else.

**Comment by koverhbarc** [comment]
I'm not sure if I should reply by e-mail or on github, so I'm posting it to both:

Yes, it does, and it also occurs on a Windows 7 machine with more memory. The actual crash happens only at the very end, after >1900 things have been deleted and SLADE has leaked 1500 MB (more memory than I've ever seen any program use before!) - that's nearly a megabyte per thing.
Just as importantly, the 3.1.2 build won't run on my XP machine, even though it is supposed to support XP (it gives the generic 'has encountered a problem and needs to close' error message on launch); the 'XP' binaries do run on 7. I transferred them over because the Win7 machine doesn't have 7zip; never needed it for anything else.

Andrew Usher

**Comment by koverhbarc** [comment]
I was able to avoid the crash on XP (not on 7) by uninstalling 3.1.0 and re-extracting 3.1.1.5 , which no longer wants a registry install (a good thing). However, this only slightly reduces the memory leak and on a map with even more Things (I used ed4 map27 - note that I had to open it as Hexen/none; as the correct Doom2/ZDoom caused a lock-up on all these configurations), I can get to the crash. It's not just Things, either, as deleting linedefs appears to consume about the same amount per deletion, but that is so slow I could not get to the crash, which occurs at about 1600 MB used (more memory than should ever be needed!).

Testing on Doom2 levels showed that I could rapidly delete everything with little memory leak, so it is non-linear, and since it is actually slightly less per Thing on the map27 than on planisf2 it must depend on the entire size of the level in memory and not just the Things portion. This gives me an idea of what might cause it (re-sorting or re-allocating memory after each deletion). To sum up your map editor just doesn't scale very well, especially in ZDoom mode.

I leave it to you to decide how urgent this is but the memory leak at least is a bug.

**Comment by sirjuddington** [comment]
Very strange, I still can't get this to happen at all - deleting all things in a map with many of them (>2000) works fine and memory usage does not go up. It may be something to do with your settings or even drivers perhaps. Would you be able to send/post your slade3.cfg file?

**Comment by koverhbarc** [comment]
I could (for XP; Windows 7 doesn't let me access it). However, these are two 'clean' installs (the .cfg file is that created by SLADE on first running) on two completely different machines, and exhibit the same behavior. I've also tried the non-XP version on 7; no difference; nor with other changes. 

More simply though: are you deleting the Things one-by-one, as I have been implying, or all together? The latter only leaks as much memory as _one_ individual deletion, about a megabyte on these levels. Nevertheless, it can be followed in Task Manager, and if you undo then mark and delete again (not 'redo') the memory usage continues to rise; no doubt it would eventually still reach the threshold and crash (without effective warning). I guess you could assume no one will do this much work in one session (closing the map editor frees all the memory) as to reach the point.

The other issue is not sudden though; large maps opened in ZDoom mode work very slow, larger ones even slower, until ed4 map27, which becomes slow enough to effectively lock up (Windows 7 kills it). zdcmp2 won't even load completely, but that's UDMF.

**Comment by sirjuddington** [comment]
Ok I think I see what you mean now (I was deleting everything in one go), each time I delete a thing in that map memory usage goes up by ~1mb

Unfortunately this is due to the way undo/redo works (favours speed over memory usage), and isn't something that would be easy to fix without completely reworking it.
